akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl.fusing
+
+import akka.stream._
+import akka.stream.stage._
+import akka.stream.scaladsl._
+import akka.stream.actor.ActorSubscriberMessage
+import akka.stream.actor.ActorSubscriberMessage._
+import akka.stream.actor.ActorPublisherMessage
+import akka.stream.actor.ActorPublisherMessage._
+import scala.concurrent.forkjoin.ThreadLocalRandom
+
+class FlattenMerge[T, M](breadth: Int) extends GraphStage[FlowShape[Source[T, M], T]] {
+  private val in = Inlet[Source[T, M]]("flatten.in")
+  private val out = Outlet[T]("flatten.out")
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(attr: Attributes) = new GraphStageLogic(shape) {
+
+    var sources = Vector.empty[LocalSource[T]]
+    var haveData = 0
+
+    val readyFilter = (src: LocalSource[T]) ⇒ src.elem != null
+    def pushOut(): Unit = {
+      val idx = ThreadLocalRandom.current.nextInt(haveData)
+      val src = sources.iterator.filter(readyFilter).drop(idx).next()
```

Comment:
Hmm, trying out a fixed-size queue triggered the thought that we might want to gracefully deal with `breadth == Int.MaxValue`, which means not allocating an 8GB or 16GB array. OTOH dropping from that iterator also becomes O(n), so a queue is better. But it needs to be dynamic, so I’m implementing that now.